integrate coredev.buildout (was PR #1670)

stevepiercy opened the pull request:
## Preview of pull request

https://plone6--1671.org.readthedocs.build/contributing/core/index.html

## Description

This time without submodules

## ToDo before merge

- ~~@jensens insert a Mermaid diagram in docs at https://github.com/plone/documentation/pull/1671/files#diff-36199f263c1ec44ec26fc136ea22e39344bf44acc065e03e0fe26bc263d2f9f2.~~ See #1683
- [x] Purge the files at the root in `coredev`.

## ToDo after merge

- [x] @plone/framework-team update links in to correct location https://plone.org/community/teams/framework
- Fix link in PLIP template `http://docs.plone.org/develop/coredev/docs/plips.html` to point to new docs.
  - [x] [Volto](https://github.com/plone/volto/blob/main/.github/ISSUE_TEMPLATE/PLIP.md?plain=1) see https://github.com/plone/volto/pull/6176
  - [x] [Products.CMFPlone](https://github.com/plone/Products.CMFPlone/blob/master/.github/ISSUE_TEMPLATE/PLIP.md?plain=1) See https://github.com/plone/Products.CMFPlone/pull/3988
  - [x] Other?

## Closes

- Closes #1495
- Closes #1486
- Closes plone/buildout.coredev#840
- Closes plone/Products.CMFPlone#3973

## References

- https://github.com/plone/documentation/issues/1278

stevepiercy committed:
Clean up PR #1670

stevepiercy committed:
Revert whitespace changes and use naked URLs (< and > are not required with a MyST extension we use), thus simplifying the diff review against `6.0`.

stevepiercy committed:
Minor grammar fixes, restore whitespace, one line per sentence, emphasis not strong

stevepiercy committed:
Revise documentation.md with current resources

stevepiercy committed:
- Set the stage for the purpose of this contributing how to guide. It's not an explanation or reference.
- Purge git and GitHub material, and refer to first-time contributors.
- Modify pre-requisites

stevepiercy committed:
Pull in RTD PR review config

stevepiercy committed:
Use correct primary branch

stevepiercy commented:
I could really use a review from anyone who cares about how to contribute to and develop in Plone core of just this one file (https://plone6--1671.org.readthedocs.build/contributing/core/index.html) in this commit: https://github.com/plone/documentation/pull/1671/commits/e2fb4f8bf4e3509e01b642b96a6bbc6b32d6a8ff @jensens @petschki @ericof @MrTango @sneridagh @davisagli @thet This one file is the key to getting new ~suckers~ contributors to Plone core.

For anyone who cares about releasing Plone, specifically @mauritsvanrees, I especially would like a review on https://plone6--1671.org.readthedocs.build/contributing/core/release.html

More to come, but these two pages are super critical to the others.

stevepiercy committed:
Merge branch '6.0' into integratel-coredev.buildout-2

petschki commented:
When it comes to "Edit packages" I always create a `local.cfg` ( -> ignored in `.gitignores` ) which looks like this:

```
[buildout]
extend = 
    core.cfg

auto-checkout +=
    plone.app.event
    icalendar
    my.package

[sources]
# add sources here, which arent in the sources.cfg already or need another branch
my.package = git git@github.com:collective/my.package.git branch=main
```

and then run 

```
$ ./bootstrap.sh -c local.cfg
```

so you do not have to edit any of the original files ( which might also get manipulated by mr.roboto ... eg. checkouts.cfg )

stevepiercy commented:
> so you do not have to edit any of the original files ( which might also get manipulated by mr.roboto ... eg. checkouts.cfg )

@thet also suggested something similar.

## `buildout.local.cfg`

```cfg
[buildout]
extends =
  buildout.cfg

auto-checkout =
    # Add packages which you want to develop
    plone.app.caching
    plone.caching
    plone.app.event
    icalendar
    # others
    ...
```

However further on, according to this ancient documentation, the release manager would look in `checkouts.cfg` to see what needs to be released. Additionally Jenkins looks at this file—and not any custom file—to run the full test suite. Is all that correct?

It would be good to verify what is the actual preferred process to contribute to Plone core.

stevepiercy committed:
bin/wsgi no workie

stevepiercy committed:
Update commit message

stevepiercy committed:
Update mrdeveloper.md

stevepiercy committed:
Update package-dependencies.md, reverting whitespace changes

stevepiercy committed:
Revert whitespace changes, restore TODO

stevepiercy committed:
Revert whitespace muckery

stevepiercy committed:
Merge branch '6.0' into integratel-coredev.buildout-2

stevepiercy committed:
Revise and make recommendations in plips.md to remove coaching and focus on how to do things.

stevepiercy committed:
Move FAQ to the end to elevate the process overview.
Revise process overview without the FWT, as it appears to be an ex-FWT, or possibly just pining for the fjords.

stevepiercy committed:
Fix linkcheck redirect

stevepiercy committed:
Revise how to submit a PLIP for today

stevepiercy committed:
Fix redirect

davisagli commented on docs/contributing/core/index.md:
````diff
@@ -0,0 +1,268 @@
+```{include} ../../volto/contributing/install-operating-system.md
+```
+
+-   Python {SUPPORTED_PYTHON_VERSIONS}
````
This is version-dependent.

Plone 6.1 will only support Python 3.10 through 3.12 (see https://community.plone.org/t/plone-6-1-will-drop-support-for-python-3-8-and-3-9/19128)

Plone 6.0 supports Python 3.8 through 3.12

Plone 5.2 supports Python 2.7 and 3.8

stevepiercy replied:
I could replace the replacement with Plone-specific versions when 6.1 is released.

- PLONE_61_SUPPORTED_PYTHON_VERSIONS
- PLONE_60_SUPPORTED_PYTHON_VERSIONS

Plone 5.2 development is documented at 5.docs.plone.org.

stevepiercy replied:
See https://github.com/plone/documentation/issues/1674 and https://github.com/plone/Products.CMFPlone/issues/3926

davisagli commented on docs/contributing/core/index.md:
```diff
@@ -0,0 +1,268 @@
+### libxml2, libxslt, and GCC
+
+Consult the resources at the links above for installation instructions.
+Package managers may facilitate installation.
```
These days on most platforms lxml can be installed as a prebuilt binary wheel that includes statically linked copies of libxml2 and libxslt. So in most cases this doesn't need to be done manually.

I think we could probably get away with removing it from the list of prerequisites here. We already don't mention it in https://plone6--1671.org.readthedocs.build/install/create-project.html

stevepiercy replied:
See ef40380271ffed73f2f45f052203193add919fdf

davisagli commented on docs/contributing/core/index.md:
```diff
@@ -0,0 +1,268 @@
+-   {term}`GNU make`
+-   {term}`Git`
+-   [libxml2 and libxslt](https://gitlab.gnome.org/GNOME/libxslt/-/releases), including development headers
+-   [GNU Compiler Collection (GCC)](https://gcc.gnu.org/) to compile {term}`ZODB`, {term}`Zope`, and {term}`lxml`
```
You need a C compiler. It can be gcc, or clang (which is what the macOS developer tools give you)

stevepiercy replied:
See ef40380271ffed73f2f45f052203193add919fdf

davisagli commented on docs/contributing/core/index.md:
````diff
@@ -0,0 +1,268 @@
+
+```shell
+./bootstrap.sh
+```
````
Note, this script currently requires python3.11 specifically to be available in user's PATH

stevepiercy replied:
There's also `.bootstrap-pyenv.sh`, which appears to be targeted toward Plone 6.1, so I omitted it.

I added a conditional step in cbb149e610f953e3f468ca2a2f844ff51b853fc0.

davisagli commented on docs/contributing/core/index.md:
```diff
@@ -0,0 +1,268 @@
+
+To visit your Plone instance, you can open the link http://0.0.0.0:8080 in a web browser.
+
+Click {guilabel}`View your Plone site` to do exactly that.
```
Running the coredev buildout installs packages and creates a Zope instance, but it doesn't create a Plone site. There's a missing step here to do that before you can view the Plone site.

stevepiercy replied:
Oops. Got ahead of myself. Fixed in 962a4b981213b3b674a6876e19e8d2d9127b01ac

davisagli commented on docs/contributing/core/index.md:
````diff
@@ -0,0 +1,268 @@
+./bin/instance fg
+```
+
+To visit your Plone instance, you can open the link http://0.0.0.0:8080 in a web browser.
````
We should clarify that this is the way to access the Classic UI. If you want to use Volto with the backend installed by the coredev buildout, then you have to:
1. Run the coredev buildout, start it, and create a Volto site.
2. Install and run Volto, and access it on a different port.

stevepiercy replied:
Just to be clear, the buildout method has links to install Volto as you suggest in Step 2.

![Screenshot 2024-06-24 at 3 11 43 AM](https://github.com/plone/documentation/assets/102112/eef337fb-d3ea-4fe1-810c-2924ca1e57f1)

[Install Plone from its packages](https://6.docs.plone.org/install/install-from-packages.html) and refer to the full Volto documentation [Frontend](https://6.docs.plone.org/volto/index.html).

Both links are incorrect.

We want to [Develop Volto core](https://6.docs.plone.org/volto/contributing/developing-core.html). But even that is not correct because the backend is in a Docker container.

So how does one install and run Volto in this context?

Until there exists a clear way to do this, I will leave Volto out of the coredev documentation.

Addressed in [edee66e](https://github.com/plone/documentation/pull/1671/commits/edee66e4fdb4df1fe47956ea57b8691db2a17887).

stevepiercy replied:
Just saw this come through the daily flood of GitHub emails:

https://github.com/plone/plone.volto/issues/155

@ericof @davisagli @sneridagh is `plone.volto`'s `README.md` the official docs for how to install Volto using buildout? The missing link, if you will.

davisagli replied:
@stevepiercy The change you made is good enough for now.

We could add a note to https://6.docs.plone.org/volto/contributing/developing-core.html#start-the-backend-and-volto explaining that you also have the option of installing and starting the Plone backend in some other way -- such as with the coredev buildout instead of `make start-backend-docker`. As long as it's running on port 8080, the default Volto settings will use it.

If you're using buildout and you're going to use the Plone backend with Volto, then you need to choose "Create a new Plone site" instead of "Create Classic UI Plone site" on the screen that looks like this:
<img width="846" alt="Screenshot 2024-06-25 at 9 46 36 PM" src="https://github.com/plone/documentation/assets/49014/3fb22b7f-0aa8-4fd8-97bb-b3389f96ea5b">
This is what installs plone.volto which applies some extra configuration needed for Volto.

This screen will be replaced with plone.distribution's install page in Plone 6.1. If you installed Plone from the cookiecutter instead of buildout, then the make commands take care of creating the Plone site for you so you don't need to use this screen.

davisagli commented on docs/contributing/core/index.md:
```diff
@@ -0,0 +1,268 @@
+The current default and development branch of `buildout.coredev` is `6.1`.
+Older versions are named according to their `major.minor` version.
+
+Always begin by checking out the git branch on which you want to work, usually called `origin`.
```
`origin` is the name of a git remote, not a branch.

stevepiercy replied:
Thanks. I think I got it technically correct now in 7e45ba71e13dd06dddcb0afd355cb28f397ce50d

davisagli commented on docs/contributing/core/index.md:
````diff
@@ -0,0 +1,268 @@
+
+```{important}
+Make sure to rerun buildout for the current branch to get the correct versions of packages, otherwise you will get some weird behavior.
+```
````
It can be awkward to try to work on different versions of Plone in the same copy of the coredev buildout, because if you switch branches most packages will need to be reinstalled on a different version. If you need to contribute to different branches of the coredev buildout, I'd recommend creating multiple clones in different folders rather than trying to switch one clone back and forth.

stevepiercy replied:
Addressed in 0897e15754a63afbc4266af801b0c56203f60ef5

davisagli commented on docs/contributing/core/index.md:
````diff
@@ -0,0 +1,268 @@
+It's a good idea to use a branch name that includes the issue number and is descriptive of what it resolves.
+
+```shell
+git switch -c 123-thing-i-fixed -t origin/6.1
````
I think we want to talk about doing this in the context of a specific package that is part of Plone core, not the coredev buildout. It's very very rare to make changes to the coredev buildout itself unless you're a release manager. Usually you make a change to the package that needs it, and then once the PR for that package is merged, mr.roboto will automatically update checkouts.cfg in the coredev buildout to record the fact that that package has changes which need to be released.

I would move discussion of creating a new branch down below "Edit packages" since you have to make sure the package is checked out before you can switch it to a new branch.

stevepiercy replied:
Addressed in 1704bbc84217f255ac9716aad06f2c0241588f89.

davisagli commented on docs/contributing/core/index.md:
````diff
@@ -0,0 +1,268 @@
+Then rerun buildout to install the source packages to edit.
+
+```shell
+./bin/buildout
````
Add the `-N` flag to save time by not checking PyPI to see if there are updates to packages that were already installed:

```suggestion
./bin/buildout -N
```

stevepiercy replied:
Done in 7e604ed4df2c597436d9c1a1e7f8a656fb1ac31f

davisagli commented on docs/contributing/core/index.md:
````diff
@@ -0,0 +1,268 @@
+
+```shell
+./bin/test -m plone.app.caching
+```
````
We could add a note that `bin/test` is a script installed by buildout which runs the Zope testrunner. Newcomers may be expecting pytest.

stevepiercy replied:
That belongs more in reference, or maybe explanation, than a how-to guide. Perhaps "Test framework of Plone core"?

davisagli commented on docs/contributing/core/index.md:
```diff
@@ -0,0 +1,268 @@
+Instead write a test that passes.
+
+After the package level tests pass with your change, you can run the full unit test suite to ensure other packages aren't affected by the change.
+It takes 5-10 minutes to run the full unit test suite.
```
You can do this, but these days we have CI, so usually I create a PR at this point and ask Jenkins to run the tests.

stevepiercy replied:
Done in fcdd9923dcb60a1007cea70e2ebf9c3bdf73544d

davisagli commented on docs/contributing/core/index.md:
```diff
@@ -0,0 +1,268 @@
+
+All changes require a change log entry.
+
+For packages that use [towncrier](https://pypi.org/project/towncrier/) to produce change logs, see {ref}`contributing-change-log-label`.
```
It might not be obvious how to tell if it uses towncrier. Look for the `news` folder.

stevepiercy replied:
Done in 7838b17d1b3a529800d30b4cae2d6cfd1c01d87c

davisagli commented on docs/contributing/core/index.md:
```diff
@@ -0,0 +1,268 @@
+New change log entries should be added at the top of {file}`CHANGES.rst`.
+
+
+## Update `checkouts.cfg`
```
Remove this section. Developers can make changes to checkouts.cfg locally in order to do development, but they should not commit them. checkouts.cfg gets updated automatically by mr.roboto after a pull request to a core package is merged.

stevepiercy replied:
Done in abf5f7b4dadf5556cd1abe8533970eb3c083bcc6

davisagli commented on docs/contributing/core/index.md:
```diff
@@ -0,0 +1,268 @@
+
+After you have completed all the foregoing steps, push your changes to a remote branch and create a pull request in GitHub.
+
+If you are working from an issue, return to the original issue, and add a link to the pull request.
```
Or mention the issue as "Fixes #issuenum" in the PR description, and github will create the link

stevepiercy replied:
Done in bc12737979848953a97e6aa2b9e18d236171b14e

davisagli commented on docs/contributing/core/index.md:
```diff
@@ -0,0 +1,268 @@
+When you push a change to a Plone package, there may be GitHub workflows that run automatically.
+The CI package [mr.roboto](https://github.com/plone/mr.roboto) will perform some checks and suggest that you run Jenkins after all other CI runs.
+
+For each Plone and Python version, Jenkins runs two jobs: one for the package itself (which will give you a fast feedback, within 10 minutes) and one on the full `coredev` build (which can take up to an hour, but makes sure no other packages are affected by your change).
```
These days Jenkins does not run a job for one specific package. For most packages there are some github actions workflows that do that, as well as some lint checks.

There are usually several jobs that run all Plone tests on Jenkins -- basically one for each combination of Plone version + Python version where that branch of the package is used, but some of them may skipped to reduce resource use (i.e. for Plone 6 we run this job on Python 3.8 and 3.11 but not 3.9 or 3.10)

stevepiercy replied:
I'm actually OK with purging this detail and not trying to document it here. It doesn't belong in a how-to guide, but more in explanation or reference.

Done in 6ef5c58872b6b68c76a8563c41a38b506ded8496

davisagli reviewed (commented):
Reviewing docs/contributing/core/index.md in particular.

stevepiercy committed:
Update pre-requisites of libx* and C compiler

stevepiercy committed:
Add admonition for non-3.11 Pythons

stevepiercy committed:
Revise and make recommendations in plips.md to remove coaching and focus on how to do things.

stevepiercy committed:
Back off from attempt to use buildout for Volto, and refer to official docs.

stevepiercy committed:
Use correct git terms, instead of gitterish

stevepiercy committed:
Give better instructions not to switch between Plone or coredev.buildout branches

stevepiercy committed:
Move new development branch under Edit packages, and use what @thet and I walked through, now that it all makes sense (for now)

stevepiercy committed:
Add -N flag option to buildout

stevepiercy committed:
Reference tests to use PR, CI, and Jenkins first, and locally only if necessary

stevepiercy committed:
Add note for how to identify towncrier repos

stevepiercy committed:
Remove obsolete section "Update `checkouts.cfg`"

stevepiercy committed:
Copy from first-time-contributors about Fixes #ISSUE-NUMBER

stevepiercy committed:
Purge obsolete explanation about what Jenkins runs

stevepiercy commented:
@davisagli thank you for your thorough feedback. Would you please take another pass at my revisions? I left several of your suggestions unresolved for you to take a closer look, and resolved trivial ones.

davisagli commented on docs/contributing/core/index.md:
```diff
@@ -0,0 +1,310 @@
+
+### C compiler
+
+You need a C compiler on your system to compile {term}`ZODB` and {term}`Zope`.
+
+On macOS, Developer Tools providers clang for a C compiler.
```
```suggestion
You need a C compiler on your system to compile some of the Python libraries that Plone uses.

On macOS, Developer Tools provides Clang for a C compiler.
```

davisagli commented on docs/contributing/core/index.md:
````diff
@@ -0,0 +1,310 @@
+The following command will switch and track changes from the remote `origin` and its branch `6.1`.
+
+```shell
+git checkout -t origin/6.1
````
This is a bad example, because 6.1 is the default branch of buildout.coredev, so it's what you already have after cloning. It would probably make more sense to show how to switch to the 6.0 branch.

stevepiercy replied:
Yeah, I tried to preserve something from 5.2, but it just doesn't make sense, now that we decided to not switch coredev branches. I rewrote this section in 26da8cc1d6fed9aa0adb12611384c29546c77e9e.

davisagli commented on docs/contributing/core/index.md:
````diff
@@ -0,0 +1,310 @@
+```
+
+```{seealso}
+For tips on working with `mr.developer`, see {doc}`mrdeveloper`.
````
```suggestion
Checking out additional repositories using the `auto-checkout` option is a feature of `mr.developer`.
For more information, see {doc}`mrdeveloper`.
```

stevepiercy replied:
Done in a80c9ccc6fb197b567e6dcc8e62d3be5164d96e0

davisagli commented on docs/contributing/core/index.md:
````diff
@@ -0,0 +1,310 @@
+It's a good idea to use a branch name that includes the issue number and is descriptive of what it resolves.
+
+```shell
+git switch -c 123-thing-i-fixed -t origin/6.1
````
```suggestion
git switch -c 123-thing-i-fixed
```

`origin/6.1` only makes sense for buildout.coredev itself, which is not where we're creating this branch. And we don't want the new branch to track an existing remote branch -- the upstream branch it should track won't exist on the remote until it is pushed.

stevepiercy replied:
Included with 26da8cc1d6fed9aa0adb12611384c29546c77e9e.

davisagli reviewed (dismissed):
@stevepiercy You did a good job incorporating my feedback! Here are a few more things from my latest pass.

stevepiercy committed:
Update C compiler pre-requisites

Co-authored-by: David Glick <david@glicksoftware.com>

stevepiercy commented:
> We could add a note to https://6.docs.plone.org/volto/contributing/developing-core.html#start-the-backend-and-volto explaining that you also have the option of installing and starting the Plone backend in some other way -- such as with the coredev buildout instead of make start-backend-docker. As long as it's running on port 8080, the default Volto settings will use it.

First, this PR is intended to migrate Plone 5.2 `buildout.coredev` docs into Plone 6 docs. Its audience is the Plone backend and Classic UI contributor. Volto already has its own contributor/developer story.

In general, when providing options to developers, we should give them sufficient information and context to make a well-informed decision. A note that says "you can do this thing another way" does not say why they should do that thing differently, and would disrupt the flow of the how-to guide.

Instead we could add an introduction to the start of https://6.docs.plone.org/volto/contributing/developing-core.html, where developers can choose from buildout or Docker-based development. However, then we would need to bring in all the pre-requisites of Volto development into this documentation, which is out of scope.

> If you're using buildout and you're going to use the Plone backend with Volto, then you need to choose "Create a new Plone site" instead of "Create Classic UI Plone site" on the screen that looks like this:

Doing so takes the contributor to a screen that has problems. See https://github.com/plone/documentation/pull/1671#discussion_r1650788403.

Volto is out of scope for this PR.

> This screen will be replaced with plone.distribution's install page in Plone 6.1.

I look forward to it.

stevepiercy committed:
Reword seealso for `mr.developer`

stevepiercy committed:
Rewrite adn simplify git pieces

stevepiercy committed:
Indicate symbol

stevepiercy committed:
Fix headings and meta information

stevepiercy committed:
Remove todo

stevepiercy committed:
- Align overview with sections of PLIP submittal process
- Add designated packages, teams, repos with a default

stevepiercy committed:
- Purge FAQ as redundant and unused. No one asks questions about PLIPs anyway, so how can they be frequent?
- Purge hand-holding. It's a how-to guide.
- Refer to setup of development environment.

stevepiercy committed:
Volto has its own contributing process, regardless of whether it is a PLIP

stevepiercy committed:
Add intersphinx configuration for Plone 5 docs

stevepiercy committed:
Finalize PLIP stuff

stevepiercy committed:
Clean up release.md and add Community Forum posts

stevepiercy commented:
@mauritsvanrees @vincentfretin would you please review the latest version of the [Plone Release Process](https://plone6--1671.org.readthedocs.build/contributing/core/release.html)?

It was last updated 4 years ago in `buildout.coredev`. I inserted a couple of points for posting to the Plone Community Forum. You might have a checklist somewhere else that we can use to update this document. Please let me know. Thank you!

stevepiercy committed:
Update troubleshooting.md

stevepiercy committed:
Reorder navigation

stevepiercy committed:
Rename troubleshooting.md to troubleshoot.md

vincentfretin commented:
@stevepiercy I didn't work with Plone since 2019. That's @erral that is taking care of plone.app.locales now, so you should probably reflect that in the Plone release process doc.

stevepiercy committed:
Replace @vincentfretin with @erral

stevepiercy commented:
Thank you for your response @vincentfretin. Updated accordingly in [cd238fe](https://github.com/plone/documentation/pull/1671/commits/cd238fe82fd6171bdb3a8361e24b31f23bf49871)

stevepiercy committed:
Add link to active PLIPs GitHub project board.

mauritsvanrees commented on docs/contributing/core/index.md:
````diff
@@ -0,0 +1,312 @@
+`which python3.11` -m venv .
+```
+
+Edit it according to the Python version you want to use, then save and close the file.
````
That will leave a local change.  Should we warn about it?

```suggestion
Edit it according to the Python version you want to use, then save and close the file.
After you have run the script you should undo the change, otherwise you have a local change in git that you might accidentally commit.
```

stevepiercy replied:
A warning would be good, but it would be even better to allow the developer to use whatever version of Python they want as an argument in the script. I'm not sure how to do that with shell scripts.

Same with the `.bootstrap-pyenv.sh` script.

mauritsvanrees commented on docs/contributing/core/index.md:
````diff
@@ -0,0 +1,312 @@
+./bootstrap.sh
+```
+
+If you run into issues in this process, see {doc}`troubleshooting`.
````
On the [preview](https://plone6--1671.org.readthedocs.build/contributing/core/index.html#install-plone-core-for-development) "troubleshooting" is no link: the `{doc}` does not seem to do anything.

stevepiercy replied:
Fixed in be8c6ad660b7c14108d719ec5ce8d7f2f1512c27

mauritsvanrees commented on docs/contributing/core/index.md:
```diff
@@ -0,0 +1,312 @@
+
+If you run acceptance tests with the `--all` option, it will repeatedly launch and close browser windows that gain focus, disrupting you from doing any other work.
+This takes 30-40 minutes to run.
+If you can't afford this disruption, you can defer it to Jenkins.
```
This does not happen for me.  I have `export ROBOT_BROWSER="headlesschrome"` in my bash profile, and the Jenkins jobs have this too.  Maybe use something like this to explain it:

```suggestion
If you run acceptance tests with the `--all` option, it will run tests in a real browser.
This takes 30-40 minutes to run.
This may repeatedly launch and close browser windows that gain focus, disrupting you from doing any other work.
If this happens, you can install the `chromedriver` OS package.
See https://developer.chrome.com/docs/chromedriver.
Then run `export ROBOT_BROWSER="headlesschrome"` and again run `bin/test --all`.
```

stevepiercy replied:
Nice! Thank you!

mauritsvanrees commented on docs/contributing/core/troubleshoot.md:
The whole `troubleshoot.md` file can be removed.  It talks about a `bootstrap.py` file.  This is no longer used in Plone coredev 5.2 or higher.  (4.3, 5.0 and 5.1 have it.)
`bootstrap.sh` is fine.  Or `bootstrap-pyenv.sh` if you must.
But if you see `bootstrap.py`: kill it with fire.  It is highly unlikely to work.

stevepiercy replied:
🔥 set in a26ee569fca6bd160ade3990b6673c35ca6d00a1.

mauritsvanrees commented on docs/contributing/core/release.md:
```diff
@@ -0,0 +1,131 @@
+
+## Release process for Plone packages
+
+To keep the Plone software stack maintainable, the Python egg release process must be automated to a high degree.
```
"egg" is an old name, we should call it "package":

```suggestion
To keep the Plone software stack maintainable, the Python package release process must be automated to a high degree.
```

mauritsvanrees commented on docs/contributing/core/release.md:
```diff
@@ -0,0 +1,131 @@
+-   {file}`README` and {file}`CHANGES` must be visible on PyPI.
+-   Released eggs must contain generated gettext `.mo` files, but these files must not be committed to the repository.
+    The `.mo` files can be created with the `zest.pocompile` add-on, which should be installed together with `zest.releaser`.
+-   `.gitignore` and `MANIFEST.in` must reflect the files going in to the egg (must include page template and `.po` files).
```
Again no "egg" please.

```suggestion
-   `.gitignore` and `MANIFEST.in` must reflect the files going in to the package (must include page template and `.po` files).
```

mauritsvanrees commented on docs/contributing/core/release.md:
```diff
@@ -0,0 +1,131 @@
+
+`plone.app.locales`
+:   Please leave this to the i18n team lead, Mikel Larreategi, `@erral` on GitHub.
+
```
Let's add a few:

```suggestion
`plone.staticresources`, `mockup`, `plonetheme.barceloneta`, `plone.classicui`:
:   Please leave this to the Classic UI team, especially Peter Mathis, Johannes Raggam and Maik Derstappen.

`plone.restapi` and `plone.volto`:
:   Please leave these to Timo Stollenwerk or David Glick.
```

mauritsvanrees reviewed (commented):
Thanks for pushing this, Steve!

I made several inline comments.

stevepiercy committed:
Update docs/contributing/core/index.md

Co-authored-by: Maurits van Rees <maurits@vanrees.org>

stevepiercy committed:
Fix doc link, grammar.

stevepiercy committed:
Use chromedriver and ROBOT_BROWSER environment variable

Co-authored-by: Maurits van Rees <maurits@vanrees.org>

stevepiercy committed:
We don't use `bootstrap.py`. Burn it with 🔥!

stevepiercy committed:
s/egg/package

Co-authored-by: Maurits van Rees <maurits@vanrees.org>

stevepiercy committed:
Add maintainers

Co-authored-by: Maurits van Rees <maurits@vanrees.org>

stevepiercy committed:
Add GitHub handles to humans

stevepiercy commented:
Thanks @mauritsvanrees! One more pass, and I think it will be good to publish.

MrTango commented on docs/contributing/core/plips.md:
```diff
@@ -0,0 +1,191 @@
+
+A Plone Improvement Proposal, or {term}`PLIP`, is a change to a Plone package that would affect everyone who uses that package.
+PLIPs go through a formal process compared to bug fixes because of their broad reach.
+The Plone Framework Team reviews all PLIPs to make sure that it's in the best interest of the broader community, and that it's of high quality.
```
This should probably say that the designated teams review their PLIPs, not the Framework team reviews all.

stevepiercy replied:
Thanks! I missed that one. I've been working on this for too long. Fixed in 004370b6e7b18932af9cba683cc45072085bf030

MrTango commented on docs/contributing/core/plips.md:
```diff
@@ -0,0 +1,191 @@
+
+| Package | Repository | Team |
+|---|---|---|
+| `Products.CMFPlone` | https://github.com/plone/Products.CMFPlone | not applicable |
```
keep @plone/framework-team for now.

stevepiercy replied:
Done in 75f782c13fde418ee429ba522429935ed82d982f

stevepiercy committed:
Restore @plone/framework-team GitHub handle

stevepiercy committed:
s/FWT/designated team

stevepiercy committed:
Update tips submodules/plone.api submodules/volto

stevepiercy committed:
Add a todo for the Steering Circle

stevepiercy committed:
Add a todo for the Steering Circle

stevepiercy committed:
Add the PLIP to the PLIP project board

stevepiercy committed:
Add the PLIP to the PLIP project board

stevepiercy committed:
Remove TODO for Steering Circle, as it is not a relevant body to review PLIPs in a timely and effective manner, given it only meets monthly as a stand-up meeting

stevepiercy committed:
Remove TODO for Steering Circle, as it is not a relevant body to review PLIPs in a timely and effective manner, given it only meets monthly as a stand-up meeting

stevepiercy committed:
Merge remote-tracking branch 'origin/integratel-coredev.buildout-2' into integratel-coredev.buildout-2

stevepiercy committed:
Revert "Update tips submodules/plone.api submodules/volto"

This reverts commit 4c16ff270dc8056e50e72979785648cf87f310df.

stevepiercy committed:
Merge branch '6.0' into integratel-coredev.buildout-2

stevepiercy committed:
Merge branch '6.0' into integratel-coredev.buildout-2

stevepiercy committed:
Merge branch '6.0' into integratel-coredev.buildout-2

stevepiercy commented:
@davisagli I would like to merge this after the Volto Team meeting on Tuesday, July 2. Will that work for you?

jensens commented:
Sorry for the delay, I'll look into it ASAP.

gforcada commented on docs/contributing/core/continuous-integration.md:
```diff
@@ -0,0 +1,111 @@
+    "keywords": "Plone, continuous integration, best practices"
+---
+
+# Essential continuous integration practices
```
This whole page is outdated, it does not mention `mr.roboto`, GHA nor `plone/meta` which make almost all the rules described outdated.

Sorry that it took me forever to react on the call to help on this PR 🙇🏾 

I can do the update, should I push it directly to this branch? 🤔

stevepiercy replied:
@gforcada please create a PR against this branch for review. Do not push commits directly to this branch.

I'm not clear on which repositories `plone/meta` applies. From what I know, many repositories use `make` and not `tox`, including Volto, plone.restapi, and Documentation. Until there is both agreement and implementation for a unified approach, I don't know whether `plone/meta` should be declared as authoritative for all projects under the Plone GitHub organization.

gforcada replied:
Ok, thanks I will do that (the branch).

As for `plone/meta` or `make`, well, it's literally all but volto related packages that already use `plone/meta`. There are a few that are missing (like CMFPlone) that I did not had the time to apply `plone/meta` to it.

But this documentation is meant for classic UI, or? 🤔 I meant I saw quite a few references that always branch to say "if you are doing volto then see ../volto"

stevepiercy replied:
This part of the  documentation is intended for anything that is not Volto, which includes Classic UI, core add-ons, and other core packages.

There is also an effort with cookiecutter and mxdev, which is intended to replace buildout, at which point another rewrite of docs will need to take place.

I don't know how that reconciles with `plone/meta`. I just document what I understand.

gforcada replied:
`plone/meta` and `mxdev`/`cookiecutter` are orthogonal, at least I'm 100% sure with `mxdev`, with `cookiecutter` that depends on what's being done, of for which purpose is used 😓 

I've seen so many `cookieXX` repositories, that I'm no longer sure what one should use 🤷🏾 

but if the goal is to replace `zc.buildout`, then `plone/meta` has nothing to do with that and should be fine to document.

Actually `plone/meta` has quite some exhaustive documentation on its own 😄

stevepiercy replied:
@gforcada even before your review and subsequent pull request, I had a lot of issues with this document. It never had a high-level review. I apologize for not bringing up my concerns earlier, but here they are now.

1. It's not clear what is the purpose of this document.
2. It's not clear who is the target audience.
3. Is it a tutorial, a how-to guide, an explanation, or a reference? See https://diataxis.fr/ for details of what these categories mean.
4. Much of the content is redundant to existing parts of the documentation, as well as within this pull request, including:
   - https://6.docs.plone.org/contributing/first-time.html#set-up-your-environment
   - https://6.docs.plone.org/contributing/index.html
   - https://github.com/plone/documentation/pull/1671/files#diff-92572efaca37b0041952183745916a031065e6fc8f04a0a5faa26ad023dc609f

Once we have an idea of what we want to do, then we can rewrite this document to achieve that goal. I'd be happy to have a chat with you in real-time in Discord, or continue the conversation here.

----

For cookie*, the primary repo is https://github.com/plone/cookieplone. It uses templates found in https://github.com/plone/cookieplone-templates. I don't know whether that repo uses `plone/meta` itself, but [its templates do](https://github.com/search?q=repo%3Aplone%2Fcookieplone-templates%20meta&type=code).

cookie* uses mxdev, as described in https://6.docs.plone.org/install/manage-add-ons-packages.html#manage-plone-backend-packages-with-mxdev.

----

Are these files `plone/meta`'s documentation?

- https://github.com/plone/meta/blob/main/README.md
- https://github.com/plone/meta/blob/main/config/README.md

gforcada replied:
In reverse order 🙃 

- yes, that's the documentation for `plone/meta`

- oh nice, thanks for the cookie pointers, now it's a bit more clear to me, the (not to be answered and solved now) is how that works with mr.bob and bobtemplates.plone and all that ecosystem, or again, we have cookie* for volto and mr.bob for classic UI? 🤔 

- for this document, I would say that it's an explanation of how the CI system works, being the target audience newcomers.

With the links you provided (thanks!) I would say it could be linked from https://6.docs.plone.org/contributing/index.html#continuous-integration, then, we could simplify my PR to remove the explanation about PRs and be more focused only on the tooling involved.

How that sounds?

As for talking in discord, sure, this week, I do have time in European evenings, something like https://www.timeanddate.com/worldclock/converter.html?iso=20240708T170000&p1=37&p2=202 ?

stevepiercy replied:
- OK on meta docs.
- I encourage you to try cookieplone for fun. If there is something missing, or is not clear how to do, then file an issue.
- I think an explanation page would be good. I do not know all the parts of CI, what each part does, or how they al work together to fulfill specified tasks.

  In that context, let's change the document title to "Continuous integration". If we keep "essential" or "practices", then we start telling the reader how to do things or what things they should do, which makes it a how-to guide, not an explanation.

  Also if you use Mermaid for diagrams, you could even draw a picture of the [workflow](https://mermaid.js.org/syntax/flowchart.html).

stevepiercy replied:
@gforcada to get this PR merged, I will remove the CI stuff entirely. It's not correct and should not be published in its current state. The rest of it is good. We can add the CI stuff in another PR.

I'll work on pulling it out of this PR, and moving it back to the coredev directory to preserve the grammar and MyST work I did. I'll do that sometime this week.

gforcada commented on docs/contributing/core/index.md:
````diff
@@ -0,0 +1,313 @@
+
+If you change the expected behavior of a feature in a package, you should write a test to cover the change.
+
+To run a test for the specific package that you modified, use the `-m` option followed by the package name, as shown in the following example.
+
+```shell
+./bin/test -m plone.app.event
````
```suggestion
To run a test for the specific package that you modified, use the `-s` option followed by the package name, as shown in the following example.

```shell
./bin/test -s plone.app.event
```

We should use `-s` as we are specifying a package, not a single module

stevepiercy committed:
Use `-s` for a package

Co-authored-by: Gil Forcada Codinachs <gil.gnome@gmail.com>

gforcada commented on docs/contributing/core/plip-review.md:
```diff
@@ -0,0 +1,110 @@
+-   Is this code maintainable?
+-   Is the code properly documented?
+-   Does the code adhere to formatting and style standards?
+-   Are there any importied deprecated modules?
```
```suggestion
-   Are there any imported deprecated modules?
```

stevepiercy committed:
Fix typo

Co-authored-by: Gil Forcada Codinachs <gil.gnome@gmail.com>

stevepiercy committed:
Merge branch '6.0' into integratel-coredev.buildout-2

tisto committed:
Merge branch '6.0' into integratel-coredev.buildout-2

tisto reviewed (dismissed):
LGTM!

stevepiercy committed:
Merge branch '6.0' into integratel-coredev.buildout-2

stevepiercy committed:
Merge branch '6.0' into integratel-coredev.buildout-2

stevepiercy committed:
Comment out continuous-integration and silence Sphinx warning

stevepiercy committed:
Purge coredev directory of now obsolete files

stevepiercy commented:
Thanks everyone! Merging!

@gforcada let's work on the CI page when you get the chance.

mauritsvanrees commented:
Thanks for pushing this, Steve!
